// @zerodev/permissions/_esm/policies/callPolicyUtils.js
import { encodeAbiParameters, getAbiItem, isHex, pad, toFunctionSelector, toHex } from "viem";
import { CALL_POLICY_CONTRACT_V0_0_1 } from "../constants.js";
import { ParamCondition } from "./types.js";
export function getPermissionFromABI({ abi, args, functionName, policyAddress }) {
    if (!abi || !functionName) {
        return {
            selector: undefined,
            rules: undefined
        };
    }
    const abiItem = getAbiItem({
        abi,
        args,
        name: functionName
    });
    if (abiItem?.type !== "function") {
        throw Error(`${functionName} not found in abi`);
    }
    const functionSelector = toFunctionSelector(abiItem);
    let paramRules = [];
    if (args && Array.isArray(args)) {
        paramRules = args
            .map((arg, i) => {
            if (!arg)
                return null;
            if (policyAddress === CALL_POLICY_CONTRACT_V0_0_1) {
                if (arg.condition === ParamCondition.ONE_OF) {
                    throw Error("The ONE_OF condition is only supported from CALL_POLICY_CONTRACT_V0_0_2 onwards. Please use CALL_POLICY_CONTRACT_V0_0_2 or a later version.");
                }
                return {
                    params: pad(isHex(arg.value)
                        ? arg.value
                        : toHex(arg.value), { size: 32 }),
                    offset: i * 32,
                    condition: arg.condition
                };
            }
            let params;
            if (arg.condition === ParamCondition.ONE_OF) {
                params = arg.value.map((value) => pad(isHex(value)
                    ? value
                    : toHex(value), { size: 32 }));
            }
            else {
                params = [
                    pad(isHex(arg.value)
                        ? arg.value
                        : toHex(arg.value), { size: 32 })
                ];
            }
            return {
                params,
                offset: i * 32,
                condition: arg.condition
            };
        })
            .filter((rule) => rule);
    }
    return {
        selector: functionSelector,
        rules: paramRules
    };
}
export const encodePermissionData = (permission, policyAddress) => {
    const permissionParam = {
        components: [
            {
                internalType: "enum CallType",
                name: "callType",
                type: "bytes1"
            },
            {
                name: "target",
                type: "address"
            },
            {
                name: "selector",
                type: "bytes4"
            },
            {
                name: "valueLimit",
                type: "uint256"
            },
            {
                components: [
                    {
                        internalType: "enum ParamCondition",
                        name: "condition",
                        type: "uint8"
                    },
                    {
                        name: "offset",
                        type: "uint64"
                    },
                    {
                        name: "params",
                        type: policyAddress === CALL_POLICY_CONTRACT_V0_0_1
                            ? "bytes32"
                            : "bytes32[]"
                    }
                ],
                name: "rules",
                type: "tuple[]"
            }
        ],
        name: "permission",
        type: "tuple[]"
    };
    const params = [permissionParam];
    const values = [permission];
    return encodeAbiParameters(params, values);
};
